indexing-service/src/main/java/org/apache/druid/indexing/common/actions/RetrieveUsedSegmentsAction.java

```diff
@@ -113,6 +136,75 @@ public TypeReference<Collection<DataSegment>> getReturnTypeReference()
 
   @Override
   public Collection<DataSegment> perform(Task task, TaskActionToolbox toolbox)
+  {
+    // When fetching segments for a datasource other than the one this task is writing to,
+    // just return all segments with the needed visibility.
+    // This is because we can't ensure that the set of returned segments is consistent throughout the task's lifecycle
+    if (!task.getDataSource().equals(dataSource)) {
+      return retrieveUsedSegments(toolbox);
+    }
+
+    final String supervisorId;
+    if (task instanceof AbstractBatchSubtask) {
+      supervisorId = ((AbstractBatchSubtask) task).getSupervisorTaskId();
+    } else {
+      supervisorId = task.getId();
+    }
+
+    final Set<ReplaceTaskLock> replaceLocksForTask = toolbox
+        .getTaskLockbox()
+        .getAllReplaceLocksForDatasource(task.getDataSource())
+        .stream()
+        .filter(lock -> supervisorId.equals(lock.getSupervisorTaskId()))
+        .collect(Collectors.toSet());
+
+    // If there are no replace locks for the task, simply fetch all visible segments for the interval
+    if (replaceLocksForTask.isEmpty()) {
+      return retrieveUsedSegments(toolbox);
+    }
+
+    Map<Interval, Map<String, Set<DataSegment>>> intervalToCreatedToSegments = new HashMap<>();
```

Comment:
I recall putting a comment here but can't find it anywhere.
I would prefer it if we moved the code here on down into a new method
`retrieveUsedSegmentsForReplace(toolbox, replaceLocks)`.